### PR TITLE
RNN 循环神经网络系列 3：编码、解码器

### LINKED ISSUE #2249: RNN 循环神经网络系列 3：编码、解码器
* 原文链接：[RECURRENT NEURAL NETWORKS (RNN) – PART 3: ENCODER-DECODER](https://theneuralperspective.com/2016/11/20/recurrent-neural-networks-rnn-part-3-encoder-decoder/)
  * Markdown文件：[文件地址](https://github.com/xitu/gold-miner/blob/master/TODO/recurrent-neural-networks-rnn-part-3-encoder-decoder.md)
  * PR 地址：#2267
  * 文章分类：**机器学习**
  ----
  * 翻译时间：2 天
  * 校对时间：1 天
  * 翻译积分：2 分
  * 校对积分：0.5 分
  * [积分有什么用](https://github.com/xitu/gold-miner/wiki)

### PR DESCRIPTION
**Pull request by changkun**
译文翻译完成，resolve #2249

### CODE REVIEW

**Comment by zcgeng** [comment]
校对认领

**Comment by fanyijihua** [comment]
@zcgeng 好的呢 :beer:

File: TODO/recurrent-neural-networks-rnn-part-3-encoder-decoder.md [hunk 1]
```diff
@@ -3,109 +3,109 @@
...
+我想创建一个非常小的数据集来使用（20 个英语和西班牙语的句子）。本教程的重点是了解如何构建一个编码解码器系统，而不是去关注这个系统对诸如机器翻译和其他 seq2seq 处理等任务的处理。所以我自己写了几个句子，然后把它们翻译成西班牙语。这就是我们的数据集。
 
-First we separate the sentences into tokens and then convert the tokens into token ids. During this process we collect a vocabulary dict and a reverse vocabulary dict to convert back and forth between tokens and token ids. For our target language (spanish), we will add an extra EOS token. Then we will pad both source and target tokens to the max length (biggest sentence in the respective datasets). This is the data we feed into our model. We use the padded source inputs as is for the encoder, but we will do further additions to the target inputs in order to get our decoder inputs and outputs.
+首先，我们将这些句子分隔为 token，然后将这些 token 转换为 token ID。在这个过程中，我们收集一个词汇表和一个反向词汇表，以便在 token 和 token ID 之间来回转换。对于我们的目标语言（西班牙语）来说，我们将添加一个额外的 EOS token。然后，我们会将源 token 和目标 token 都填充到（对应数据集中最长句子的）最大长度。这是我们模型的输入数据。对于编码器而言，我们将填充后的源内容直接进行输入，而对于目标内容做进一步处理，以获得我们的解码器输入和输出。
```
Comment (zcgeng) [inline TODO/recurrent-neural-networks-rnn-part-3-encoder-decoder.md]: 「词汇表」=>「词汇字典」
「反向词汇表」=>「反向词汇字典」

File: TODO/recurrent-neural-networks-rnn-part-3-encoder-decoder.md [hunk 2]
```diff
@@ -3,109 +3,109 @@
...
+    # RNN 解码器的输出
     self.all_decoder_outputs, self.decoder_state = tf.nn.dynamic_rnn(
         cell=self.decoder_stacked_cell,
         inputs=self.embedded_decoder_inputs,
         sequence_length=self.sp_seq_lens, time_major=False,
         initial_state=self.decoder_initial_state)
 ```
 
-But what about the paddings, they will also be predicting some output target but we don’t really care about those but they will still impact our loss if we factor them in. Here’s where we will be masking the loss to remove influence from paddings in the targets.
+那填充值会发生什么呢？它们也会预测一些输出目标，而我们并不关心这些内容，但如果我们把它们考虑进去，它们仍然会影响我们的损失函数。接下来我们将在屏蔽掉这些损失以消除对目标结果的影响。
```
Comment (zcgeng) [inline TODO/recurrent-neural-networks-rnn-part-3-encoder-decoder.md]: 「接下来我们将在屏蔽掉这些损失以消除对目标结果的影响。」
=>
「接下来我们将屏蔽掉这些损失以消除对目标结果的影响。」

File: TODO/recurrent-neural-networks-rnn-part-3-encoder-decoder.md [hunk 3]
```diff
@@ -3,109 +3,109 @@
...
 
-## Loss Masking:
+## 损失屏蔽
 
-We will use the targets and where ever the target is a PAD, we will mask the loss for that location to 0\. So when we get to the last relevant decoder token, the appropriate target will be an EOS token id. For the next decoder input the target will be a PAD id. This is where the masking starts.
+我们要使用那些填充的目标，并将这个位置的损失屏蔽为 0。因此，当我们获得最后一个有关的解码器 token 时，目标就会是表示 EOS 的 token ID。而对于下一个解码器的输入而言，输入目标则会是一个填充物，这也就是屏蔽开始的地方。
```
Comment (zcgeng) [inline TODO/recurrent-neural-networks-rnn-part-3-encoder-decoder.md]: 「我们要使用那些填充的目标，并将这个位置的损失屏蔽为 0。」=>
「我们会检查目标，并将目标中被填充的部分屏蔽为 0。」

「而对于下一个解码器的输入而言，输入目标则会是一个填充物」 =>
「而对于下一个解码器的输入而言，目标就会是 PAD ID」

File: TODO/recurrent-neural-networks-rnn-part-3-encoder-decoder.md [hunk 4]
```diff
@@ -116,22 +116,21 @@ self.loss = tf.reduce_mean(
     tf.reduce_sum(masked_losses, reduction_indices=1))
 ```
 
-We will cleverly use the fact that PAD IDs are 0 to apply the loss mask. Once we apply the mask, we just compute the sum of the losses for each row (sample in the batch) and then take the mean of all the sample’s losses to get the batch’s loss. From here, we can just train on minimizing this loss.
+注意到可以使用填充物的 ID 为 0 这个事实作为屏蔽手段，我们便只需计算（一个批次中样本的）每一行损失之和即可，然后取所有样本损失的平均值，从而得到一个批次的损失。这时，我们可以开始训练了。
```
Comment (zcgeng) [inline TODO/recurrent-neural-networks-rnn-part-3-encoder-decoder.md]: 「填充物的 ID 」=>「PAD ID」
「这时，我们可以开始训练了。」=>「这时，我们就可以通过最小化这个损失函数来进行训练了。」

File: TODO/recurrent-neural-networks-rnn-part-3-encoder-decoder.md [hunk 5]
```diff
@@ -3,109 +3,109 @@
...
+2. [RNN 循环神经网络系列 2：文本分类](https://github.com/xitu/gold-miner/blob/master/TODO/recurrent-neural-networks-rnn-part-2-text-classification.md)
+3. [RNN 循环神经网络系列 3：编码、解码器](https://github.com/xitu/gold-miner/blob/master/TODO/recurrent-neural-networks-rnn-part-3-encoder-decoder.md)
+4. [RNN 循环神经网络系列 4：ATTENTIONAL INTERFACES](https://github.com/xitu/gold-miner/blob/master/TODO/recurrent-neural-network-rnn-part-4-attentional-interfaces.md)
+5. [RNN 循环神经网络系列 5：CUSTOM CELLS](https://github.com/xitu/gold-miner/blob/master/TODO/recurrent-neural-network-rnn-part-5-custom-cells.md)
 
-# RECURRENT NEURAL NETWORKS (RNN) – PART 3: ENCODER-DECODER
+# # RNN 循环神经网络系列 3：编码、解码器
 
-In this post, I will cover the basic encoder-decoder which we use to process seq-seq tasks such as machine translation. We will not be covering attention in this post but we will implement it in the next one.
+在本文中，我将介绍基本的编码器（encoder）和解码器（decoder），用于处理诸如机器翻译之类的 seq2seq 任务。我们不会在这篇文章中介绍 Attention 模型，而在下一篇文章中去实现它。
```
Comment (zcgeng) [inline TODO/recurrent-neural-networks-rnn-part-3-encoder-decoder.md]: 「Attention 模型」=>「注意力模型」

File: TODO/recurrent-neural-networks-rnn-part-3-encoder-decoder.md [hunk 6]
```diff
@@ -116,22 +116,21 @@ self.loss = tf.reduce_mean(
...
 
-Here are the training results:
+以下是训练结果：
 
 ![Screen Shot 2016-11-19 at 4.56.18 PM.png](https://theneuralperspective.files.wordpress.com/2016/11/screen-shot-2016-11-19-at-4-56-18-pm.png?w=620)
 
-We won’t be doing any inference here but you can find it the following post with attention. But if you really want to implement inference here, just use the same model as training but you need to feed the predicted target back in as an input for the next decoder rnn cell. You need to embed with the same set of weights used to embed INTO the decoder and have it as another input to the rnn. This means that for the initial GO token, you need to feed in some dummy input token that will be embedded.
-
-## Conclusion:
+我们不会在这里做任何的模型推断，但是你可以在接下来的关于 Attention 模型的文章中看到。如果你真的想在这里实现模型推断，使用相同的模型就可以了，但你还得将预测目标的结果作为输入接入下一个 RNN 解码器单元。同时你还要将相同的权重集嵌入解码器中，并将其作为 RNN 的另一个输入。这意味着对于初始的 GO token 而言，你得嵌入一些伪造的 token 进行输入。
```
Comment (zcgeng) [inline TODO/recurrent-neural-networks-rnn-part-3-encoder-decoder.md]: 「Attention 模型」=>「注意力模型」

**Comment by leviding** [comment]
已经 merge 啦~ 快快麻溜发布到掘金专栏然后给我发下链接，方便及时添加积分哟。

掘金翻译计划有自己的知乎专栏，你也可以投稿哈，推荐使用一个[好用的插件](https://chrome.google.com/webstore/detail/minimalist-markdown-edito/pghodfjepegmciihfhdipmimghiakcjf?hl=zh-CN)。
专栏地址：https://zhuanlan.zhihu.com/juejinfanyi

**Comment by changkun** [comment]
链接：https://juejin.im/post/59fc1616f265da432b4a2d44
